File: SinaWeibo/ViewControllers/MyInfoViewController.m
//
//  MyInfoViewController.m
//  SinaWeibo
//
//  Created by cxjwin on 13-9-27.
//  Copyright (c) 2013年 cxjwin. All rights reserved.
//

#import "MyInfoViewController.h"
#import "SinaWeiboManager.h"
#import "WeiboUser+Builder.h"
#import "WeiboStatus+Builder.h"
#import "StatusCell.h"

#define kAvatarViewTag 201
const NSUInteger kNumberOfSections = 5;

@interface MyInfoViewController ()

@property (strong, nonatomic) WeiboStatus *latestStatus;

@end

@implementation MyInfoViewController
{
    __weak WeiboUser *currentUer;
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
    if (self) {
        // Custom initialization
    }
    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self name:kWeiboUserInfoDidUpdateNotification object:nil];
}

- (void)viewDidLoad
{
    [super viewDidLoad];
	// Do any additional setup after loading the view.
    self.tableView.backgroundColor = [UIColor colorWithPatternImage:[UIImage imageNamed:@"BackgroundTile"]];
    
    currentUer = [[SinaWeiboManager defaultManager] user];
    self.latestStatus = [[WeiboStatus alloc] init];
    
    [self createTableHeaderView];
    [self createTableFooterView];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(refreshUserView) name:kWeiboUserInfoDidUpdateNotification object:nil];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    if (currentUer.userId) {
        [self refreshUserView];
    } else {
        [[SinaWeiboManager defaultManager] getCurrentUserInfo];
    }
}

- (void)didReceiveMemoryWarning
{
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)createTableHeaderView
{
    CGRect frame = CGRectMake(0, 0, CGRectGetWidth(self.tableView.frame), 80);
    UIView *view = [[UIView alloc] initWithFrame:frame];
    
    CGRect imageViewFrame = CGRectMake(0, 0, 56, 56);
    UIImageView *imageView = [[UIImageView alloc] initWithFrame:imageViewFrame];
    imageView.center = CGPointMake(40, 40);
    imageView.tag = kAvatarViewTag;
    [view addSubview:imageView];
    
    self.tableView.tableHeaderView = view;
}

- (void)createTableFooterView
{
    
}

- (void)refreshUserView
{
    __weak UIImageView *imageView = (id)[self.tableView.tableHeaderView viewWithTag:kAvatarViewTag];
    
    SinaWeiboCommunicator *communicator = [[SinaWeiboManager defaultManager] communicator];
    NSURLCache *URLCache = communicator.defaultSession.configuration.URLCache;
    
    NSURL *avatarUrl = nil;
    if (currentUer.profileImageUrl) {
        if (currentUer.avatarLarge) {
            avatarUrl = [NSURL URLWithString:currentUer.avatarLarge];
        } else {
            avatarUrl = [NSURL URLWithString:currentUer.profileImageUrl];
        }
    }
    if (avatarUrl) {
        NSURLRequest *request = [NSURLRequest requestWithURL:avatarUrl];
        NSCachedURLResponse *response = [URLCache cachedResponseForRequest:request];
        if (response) {
            dispatch_async(dispatch_get_main_queue(), ^{
                imageView.image = [UIImage imageWithData:response.data];
            });
        } else {
            [communicator downloadImageWithURL:avatarUrl 
                     downloadCompletionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                         if (error == nil && [data length] > 0) {
                             dispatch_async(dispatch_get_main_queue(), ^{
                                 imageView.image = [UIImage imageWithData:data];
                             });
                         }
                     }];
        }
    }
    
    [self.latestStatus fillInDetailsWithJSONObject:currentUer.status];
    [self.tableView reloadData];
}

#pragma mark - 
#pragma mark - UITableViewDataSource
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;//kNumberOfSections;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 1;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *indentifier_1 = @"StatusCell";
    UITableViewCell *cell = nil;
    switch (indexPath.section) {
        case 0:
        {
            StatusCell *_cell = (StatusCell *)[tableView dequeueReusableCellWithIdentifier:indentifier_1];
            _cell.statusView.status = self.latestStatus;
            cell = _cell;
        }
            break;
            
        default:
            break;
    }
    
    return cell;
}

- (BOOL)hasPictureInStatus:(WeiboStatus *)status {
    WeiboStatus *picStatus = (status.retweetedStatus != nil ? status.retweetedStatus : status);
    if ([picStatus.picUrls count] > 0 || picStatus.thumbnailPic || picStatus.bmiddlePic || picStatus.originalPic) {
        return YES;
    } else {
        return NO;
    }
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    CGFloat height = 0;
    switch (indexPath.section) {
        case 0:
        {
            // height for avatar & name & from ...
            height = 48;
            
            // height for status text
            height = 3 + height + [self.latestStatus.contentTextSize CGSizeValue].height + 3;// status text
            if (self.latestStatus.retweetedStatus) {
                // height for separtion view
                height = 2 + height + 16 + 2;
                // height for retweeted status text
                height = 3 + height + [self.latestStatus.retweetedStatus.contentTextSize CGSizeValue].height + 3;
            }
            
            if ([self hasPictureInStatus:self.latestStatus]) {
                // height for pic view
                height = 3 + height + 64 + 3;
            }
            
            // height for tool bar
            height = 5 + height + 16 + 10;
        }
            break;
            
        default:
            break;
    }
    
    return height;
}

@end

File: SinaWeibo/ViewControllers/MyInfoViewController.swift
//
//  MyInfoViewController.swift
//  SinaWeibo
//
//  Created by cxjwin on 6/13/14.
//  Copyright (c) 2014 cxjwin. All rights reserved.
//

import UIKit

let kMyAvatarViewTag = 101

class MyInfoViewController: UITableViewController {

    override func viewDidLoad() {
        super.viewDidLoad()

        // Uncomment the following line to preserve selection between presentations
        // self.clearsSelectionOnViewWillAppear = false

        // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
        // self.navigationItem.rightBarButtonItem = self.editButtonItem
    }
	
	override func viewWillAppear(animated: Bool) {
		super.viewWillAppear(animated)
		
		let communicator = SinaWeiboManager.defaultManager().communicator
		let curUser = SinaWeiboManager.defaultManager().user

		var imageView = self.tableView.tableHeaderView.viewWithTag(kMyAvatarViewTag) as UIImageView
		
		var avatarUrl: NSURL?
		
		if curUser.profileImageUrl {
			if curUser.avatarLarge {
				avatarUrl = NSURL(string: curUser.avatarLarge)
			} else {
				avatarUrl = NSURL(string: curUser.profileImageUrl)
			}
		}
		
		if var url = avatarUrl {
			var request = NSURLRequest(URL: url)
			
			var URLCache: NSURLCache? = communicator.defaultSession.configuration.URLCache
			
			var response:NSCachedURLResponse? = URLCache?.cachedResponseForRequest(request)
			
			if response {
				var data:NSData? = response!.data
				if data {
					var image = UIImage(data: data)
					println("\(image)")
					imageView.image = image
				}
			} else {
				communicator.downloadImageWithURL(url, downloadCompletionHandler: {
					[weak imageView] (data: NSData?, _, _) in
					if data?.length > 0 {
						NSOperationQueue.mainQueue().addOperationWithBlock{
							var image = UIImage(data: data)
							println("\(image)")
							imageView!.image = image
						}
					}
					})
			}
		}
	}

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    // #pragma mark - Table view data source

    override func numberOfSectionsInTableView(tableView: UITableView?) -> Int {
        // #warning Potentially incomplete method implementation.
        // Return the number of sections.
        return 0
    }

    override func tableView(tableView: UITableView?, numberOfRowsInSection section: Int) -> Int {
        // #warning Incomplete method implementation.
        // Return the number of rows in the section.
        return 0
    }

    /*
    override func tableView(tableView: UITableView?, cellForRowAtIndexPath indexPath: NSIndexPath?) -> UITableViewCell? {
        let cell = tableView.dequeueReusableCellWithIdentifier("reuseIdentifier", forIndexPath: indexPath)

        // Configure the cell...

        return cell
    }
    */

    /*
    // Override to support conditional editing of the table view.
    override func tableView(tableView: UITableView?, canEditRowAtIndexPath indexPath: NSIndexPath?) -> Bool {
        // Return NO if you do not want the specified item to be editable.
        return true
    }
    */

    /*
    // Override to support editing the table view.
    override func tableView(tableView: UITableView?, commitEditingStyle editingStyle: UITableViewCellEditingStyle, forRowAtIndexPath indexPath: NSIndexPath?) {
        if editingStyle == .Delete {
            // Delete the row from the data source
            tableView.deleteRowsAtIndexPaths([indexPath], withRowAnimation: .Fade)
        } else if editingStyle == .Insert {
            // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
        }    
    }
    */

    /*
    // Override to support rearranging the table view.
    override func tableView(tableView: UITableView?, moveRowAtIndexPath fromIndexPath: NSIndexPath?, toIndexPath: NSIndexPath?) {

    }
    */

    /*
    // Override to support conditional rearranging of the table view.
    override func tableView(tableView: UITableView?, canMoveRowAtIndexPath indexPath: NSIndexPath?) -> Bool {
        // Return NO if you do not want the item to be re-orderable.
        return true
    }
    */

    /*
    // #pragma mark - Navigation

    // In a storyboard-based application, you will often want to do a little preparation before navigation
    override func prepareForSegue(segue: UIStoryboardSegue?, sender: AnyObject?) {
        // Get the new view controller using [segue destinationViewController].
        // Pass the selected object to the new view controller.
    }
    */

}
